### ISSUE TITLE
Controller > Service > MediatR > Repository.... or Controller > MediatR > Service > ... ?

### ISSUE THREAD

**Issue by codefactorydevelopment** [jbogard/MediatR#614]
I find both in many examples over web but not definitive answer why.

It seems to me that having `Controller > Service > MediatR > Repository` is the natural way especially when it comes to the situation that POST/PUT/PATCH (in other words, command http requests) can return results because within the service we can issue a command and after that issue a query in order to return result to the client. Having MediatR directly in controller would either force command handler to return value or have retrieve logic in controller, both of which is not nice.

If not done that way, i see no other way how to return something to the client in response to POST requests and not violate CQRS rules that commands don't return anything.

In short:
Is it usual/best practice to have `Controller > Service > MediatR > ...`?

thanks

**Comment by jbogard** [comment]
I do `Controller > MediatR > whatever`. Encapsulate the ENTIRE operation in your command handler. Don't separate those into "command" and "query". Make your command handler also return the result needed for the response.

Commands can return a result/response, as you can see in the HTTP standards. Just make your command handler return the response.

As to what happens inside your handler to perform the operation and return a response, I start with the simplest thing that could possibly work, just procedural code inside your handler. Then you can use code smells to refactor from there.

**Comment by codefactorydevelopment** [comment]
@jbogard 
thanks for your time & effort to respond, really appreciate it!

wouldn't that turn command handlers into service, actually and go against the whole concept? because, one clients case would need one result, the other would want other for the same command (`PUT someObject`, e.g.).

I must admint that that statement "_Don't separate those into "command" and "query_" sounds very odd to me especially on CQRS related library. I might very well be missing something but could you please elaborate a bit more on that part: why in particular is `Controller > Service > MediatR > ....` not good practice?

**Comment by ardalis** [comment]
Jimmy's approach of Keep-It-Simple is a good place to start. Just use `Controller > MediatR > do work` as a default approach. If you desire more structure and consistency between how various handlers behave, or if you don't always need to do much interesting work (many queries are trivial for instance), you can look at migrating from Controllers to ApiEndpoints and then only using MediatR when there is a reason to use its pipeline. Most queries will just be `Endpoint -> Repository` or even `Endpoint -> DbContext`.

See [ApiEndpoints ](https://github.com/ardalis/ApiEndpoints)if you're not familiar with it already.

**Comment by codefactorydevelopment** [comment]
@ardalis  thanks for your response!

i believe controllers vs endpoints is a bit orthogonal to this subject. i agree on everything you say about endpoints (i am familiar with them) but i'm not sure how that answers the question/dilemma.

dilemma is: doesn't merging command and query into same handler method violate SRP and defeats the purpose of CQRS?
and, a broader question, is `Controller > Service > MediatR > ,,,` considered a bad practice and if so - why?

I'd really like to hear your take on it.
thanks!

**Comment by ardalis** [comment]
I think it (endpoints vs. controllers) is related because it's often the reason teams reach for MediatR in the first place - to try to get a handle on their controllers. I see a lot of teams who go that route and end up with Controllers that have a bunch of Actions but each one is just a call to `await _mediator.Send(model);` which is a lot of boilerplate, repetitive code. If that's most of what a team is using MedaitR for, ApiEndpoints may be useful.

But to answer your question. I don't typically call out to MediatR from services. My logic flow is typically one of:
`UI > AppService > Data Abstraction > Data`
or
`UI > Data Abstraction > Data`

I don't use an `AppService` layer dogmatically if it's not adding any value. Most apps have simple endpoints for things like `GetById` that don't benefit from routing through a separate service layer. I'll just use a repo and let the endpoint be responsible for mapping the result, which still results in an endpoint with < 5 lines of code, typically.

I do typically keep an abstraction around my data because it lets me perform interception easily via DI and it lets me ensure my UI layer and my Core Domain layer have no direct dependency on volatile dependencies that I keep in a separate Infrastructure project. Jimmy and I disagree on this one but I find it's an easy way to help ensure developers don't (accidentally or intentionally) hardwire the UI to the database while adding some business logic and produce a mess of spaghetti code leading to a big ball of mud architecture. I prefer [Clean Architecture](https://github.com/ardalis/CleanArchitecture) which helps enforce [Separation of Concerns](https://deviq.com/principles/separation-of-concerns) at compile time via project references.

HTH and sorry Jimmy for hijacking your repo's issue... :)

**Comment by jbogard** [comment]
A large part of this comes down to "What conceptual level do you apply CQRS"? I apply it to the sort of "application service" level, as I find that it provides the most universal value. Not say, to "every incidence of calling a `DbContext` needs to be wrapped in a command/query".

**Comment by codefactorydevelopment** [comment]
@ardalis @jbogard 

guys, your responses gave me some very valuable insights and thank you both a bunch!

as far as i'm concerned, this issue can be closed.
